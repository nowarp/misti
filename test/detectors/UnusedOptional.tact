fun f(): Int { return 42; }
fun fOpt(): Int? { return null; }

fun use(_a: Int?) {}

message M {
    field: Int?;
}

trait T {
}

contract C with T {
    // Bad: Unused field
    _f1: Int? = 12;

    // OK: Fields are used
    _f2: Int? = 12;
    _f3: Int? = 12;
    _f4: Int? = 12;

    // OK: Don't report optionals coming from structure/message fields, since
    // we might not have information about their types in the detector.
    receive(m: M) {
        let _a: Int? = m.field;
        self._f4 = m.field;
    }

    receive("whatever") {
        // Bad: checking that the detector works in receivers
        let _a: Int? = 12;
    }

    fun test() {
        // OK: We don't report values resulting from function calls
        let _a1: Int? = f();
        let _a2: Int? = fOpt();
        let _a3: Int? = self.m();
        let _a4: Int? = self.mOpt();

        // Bad: unused optionals will be reported
        let _a5: Int? = 12;
        let _a6: Int? = 12;
        _a6!!;

        // OK: Optional is used
        let _a7: Int? = 12;
        if (_a7 == null) {}
        let _a8: Int? = 12;
        if (null == _a8) {}
        let _a9: Int? = 12;
        use(_a9);

        // OK: Field is unused, but we intentionally skip field access operations
        let _a10: Int = self._f1!!;

        // OK: Used fields
        use(self._f2);
        if (self._f3 == null) {}
    }

    fun m(): Int { return 42; }
    fun mOpt(): Int? { return null; }
}
