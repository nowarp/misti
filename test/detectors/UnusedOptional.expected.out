[LOW] UnusedOptional: Unused optional modifier: _f1
test/detectors/UnusedOptional.tact:15:5:
  14 |     // Bad: Unused field
> 15 |     _f1: Int? = 12;
           ^
  16 | 
Help: Remove optional modifier: `_f1: Int`
See: https://nowarp.io/tools/misti/docs/detectors/UnusedOptional

[LOW] UnusedOptional: Unused optional modifier: _a
test/detectors/UnusedOptional.tact:31:13:
  30 |         // Bad: checking that the detector works in receivers
> 31 |         let _a: Int? = 12;
                   ^
  32 |     }
Help: Remove optional modifier: `let _a: Int`
See: https://nowarp.io/tools/misti/docs/detectors/UnusedOptional

[LOW] UnusedOptional: Unused optional modifier: _a5
test/detectors/UnusedOptional.tact:42:13:
  41 |         // Bad: unused optionals will be reported
> 42 |         let _a5: Int? = 12;
                   ^
  43 |         let _a6: Int? = 12;
Help: Remove optional modifier: `let _a5: Int`
See: https://nowarp.io/tools/misti/docs/detectors/UnusedOptional

[LOW] UnusedOptional: Unused optional modifier: _a6
test/detectors/UnusedOptional.tact:43:13:
  42 |         let _a5: Int? = 12;
> 43 |         let _a6: Int? = 12;
                   ^
  44 |         _a6!!;
Help: Remove optional modifier: `let _a6: Int`
See: https://nowarp.io/tools/misti/docs/detectors/UnusedOptional